src/loader/tsbgw_message_queue.c

```diff
@@ -0,0 +1,273 @@
+#include <postgres.h>
+
+#include <miscadmin.h>
+#include <storage/lwlock.h>
+#include <storage/shmem.h>
+#include <storage/proc.h>
+#include <storage/procarray.h>
+#include <storage/shm_mq.h>
+#include <access/xact.h>
+
+#include "../compat.h"
+
+#include "tsbgw_message_queue.h"
+
+#define TSBGW_MAX_MESSAGES 16
+#define TSBGW_MESSAGE_QUEUE_NAME "timescale_bgw_message_queue"
+#define TSBGW_MQ_TRANCHE_NAME "timescale_bgw_mq_tranche"
+
+#define TSBGW_ACK_QUEUE_SIZE (MAXALIGN(shm_mq_minimum_size + sizeof(int)))
+typedef struct TsbgwMessageQueue
+{
+	pid_t		reader_pid;		/* should only be set once at cluster launcher
+								 * startup */
+	LWLock	   *lock;			/* pointer to the lock to control
+								 * adding/removing elements from queue */
+	uint8		read_upto;
+	uint8		num_elements;
+	TsbgwMessage buffer[TSBGW_MAX_MESSAGES];
+}			TsbgwMessageQueue;
+
+typedef enum QueueResponseType
+{
+	MESSAGE_SENT = 0,
+	QUEUE_FULL,
+	READER_DETACHED
+}			QueueResponseType;
+
+static TsbgwMessageQueue * tsbgw_mq = NULL;
+
+static void
+queue_init(bool reinit)
+{
+	bool		found;
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+	tsbgw_mq = ShmemInitStruct(TSBGW_MESSAGE_QUEUE_NAME, sizeof(TsbgwMessageQueue), &found);
+	if (!found || reinit)
+	{
+		memset(tsbgw_mq, 0, sizeof(TsbgwMessageQueue));
+		tsbgw_mq->reader_pid = InvalidPid;
+		tsbgw_mq->lock = &(GetNamedLWLockTranche(TSBGW_MQ_TRANCHE_NAME))->lock;
+	}
+	LWLockRelease(AddinShmemInitLock);
+
+}
+
+
+/* this gets called when shared memory is initialized in a backend (shmem_startup_hook) */
+extern void
+tsbgw_message_queue_shmem_startup(void)
+{
+	queue_init(false);
+}
+
+/* this is called in the loader during server startup to allocate a shared memory segment*/
+extern void
+tsbgw_message_queue_alloc(void)
+{
+	RequestAddinShmemSpace(sizeof(TsbgwMessageQueue));
+	RequestNamedLWLockTranche(TSBGW_MQ_TRANCHE_NAME, 1);
+}
+
+
+/* Notes on managing the queue/locking:
+ * We decided that for this application, simplicity of locking scheme was more important than
+ * being very good about concurrency as the frequency of these messages will be low and the
+ * number of messages on this queue should be low, given that they mostly happen when we update
+ * the extension. Therefore we decided to simply take an exclusive lock whenever we were modifying
+ * anything in the shared memory segment to avoid collisions.
+ */
+
+/* Add a message to the queue - we can do this if the queue is not full */
+static QueueResponseType queue_add(TsbgwMessageQueue * queue, TsbgwMessage * message)
```

Comment:
`static QueueResponseType` should be on its own line. Shouldn't this fixed by `pgindent`? Also, extra whitespaces around `*`

Same for functions below.